tensorboard/webapp/plugins/npmi/util/coordinate_data.ts

```diff
@@ -0,0 +1,62 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {ValueData, AnnotationDataListing} from '../store/npmi_types';
+
+export type Coordinate = {
+  runId: string;
+  annotation: string;
+  values: ValueData[];
+};
+
+export function coordinateData(
+  annotationData: AnnotationDataListing,
+  selectedAnnotations: string[],
+  activeRuns: string[],
+  activeMetrics: string[]
+): {coordinates: Coordinate[]; extremes: {min: number; max: number}} {
+  const result: Coordinate[] = [];
+  const extremeValues = {max: -1.0, min: 1.0};
+  selectedAnnotations.forEach((annotation) => {
+    const data = annotationData[annotation];
+    const allRuns = new Set(activeRuns);
+    const allMetrics = new Set(activeMetrics);
```

Comment:
move these out of the forEach since it does not need to be recreated over and over.